geth/node/peerpool.go

```diff
@@ -0,0 +1,291 @@
+			}
+		}
+	}
+}
+
+// processFoundNode called when node is discovered by kademlia search query
+// 2 important conditions
+// 1. every time when node is processed we need to update discoveredTime and reset dropped boolean.
+//    peer will be considered as valid later only if it was discovered < 90s ago and wasn't dropped recently
+// 2. if peer is connected or if max limit is reached we are not a adding peer to p2p server
+func (p *PeerPool) processFoundNode(server *p2p.Server, currentlyConnected int, topic discv5.Topic, node *discv5.Node) (connected bool) {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+	peersTable := p.peers[topic]
+	limits := p.config[topic]
+	if info, exist := peersTable[node.ID]; exist {
+		info.discoveredTime = mclock.Now()
+		info.dropped = false
+	} else {
+		peersTable[node.ID] = &peerInfo{
+			connected:      false,
+			dropped:        false,
+			discoveredTime: mclock.Now(),
+			node: discover.NewNode(
+				discover.NodeID(node.ID),
+				node.IP, node.UDP, node.TCP),
+		}
+	}
+	if currentlyConnected < limits[1] && !peersTable[node.ID].connected {
+		log.Debug("peer connected", "ID", node.ID, "topic", topic)
+		server.AddPeer(peersTable[node.ID].node)
+		peersTable[node.ID].connected = true
+		connected = true
+		if p.cache != nil {
+			if err := p.cache.AddPeer(node, topic); err != nil {
+				log.Error("failed to persist a peer", "error", err)
+			}
+		}
+	}
+	return connected
+}
+
+// processDisconnectedNode is called when node was dropped by p2p server
+// - removes a peer, cause p2p server now relies on peer pool to maintain required connections number
+// - if there is a valid peer in peer table add it to a p2p server
+//   peer is valid if it wasn't dropped recently, it is not stale (90s) and not currently connected
+func (p *PeerPool) processDisconnectedNode(server *p2p.Server, topic discv5.Topic, nodeID discover.NodeID) (connected bool) {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+	peersTable := p.peers[topic]
+	// either inbound or connected from another topic
+	if _, exist := peersTable[discv5.NodeID(nodeID)]; !exist {
+		return true
+	}
+	node := peersTable[discv5.NodeID(nodeID)].node
+	server.RemovePeer(node)
+	if p.cache != nil {
+		if err := p.cache.RemovePeer(discv5.NodeID(nodeID), topic); err != nil {
+			log.Error("failed to remove peer from cache", "error", err)
+		}
+	}
+
+	// TODO use a heap queue and always get a peer that was discovered recently
+	for _, info := range peersTable {
+		if !info.connected && !info.dropped && mclock.Now() < info.discoveredTime+mclock.AbsTime(foundTimeout) {
+			log.Debug("adding peer from pool", "ID", info.node.ID, "topic", topic)
+			if p.cache != nil {
+				// this conversions are getting funny
+				if err := p.cache.AddPeer(discv5.NewNode(
+					discv5.NodeID(info.node.ID),
+					info.node.IP, info.node.UDP, info.node.TCP), topic); err != nil {
+					log.Error("failed to persist a peer", "error", err)
+				}
+			}
+			server.AddPeer(info.node)
```

Comment:
code around that looks very similar to L213.
consider extracting it to a method